lib/experimental/interactivity-api/blocks.php

```diff
@@ -6,6 +6,29 @@
  * @package gutenberg
  */
 
+/**
+ * Adds Interactivity API directives to the File block markup using the Tag Processor.
+ *
+ * @param string   $block_content Markup of the File block.
+ * @param array    $block         The full block, including name and attributes.
+ * @param WP_Block $instance      The block instance.
+ *
+ * @return string File block markup with the directives injected when applicable.
+ */
+function gutenberg_block_core_file_add_directives_to_content( $block_content, $block, $instance ) {
+	if ( empty( $instance->attributes['displayPreview'] ) ) {
+		return $block_content;
+	}
+	$processor = new WP_HTML_Tag_Processor( $block_content );
+	$processor->next_tag();
+	$processor->set_attribute( 'data-wp-island', '' );
+	$processor->next_tag( 'object' );
+	$processor->set_attribute( 'data-wp-init', 'effects.core.file.init' );
+	$processor->set_attribute( 'hidden', true );
```

Comment:
It feels a bit strange to me to use a function to mutate the DOM when we have a primitive to do so. Have you considered using `data-wp-bind` instead? Something like this:

```php
$processor->set_attribute( 'data-wp-bind.hidden', 'selectors.core.browserSupportsPdfs' );
```

```js
import { browserSupportsPdfs } from "./utils";

store({ selectors: { core: { browserSupportsPdfs } } });
```

Reply:
Will `bind` run only once on the client? I wasn't sure how it would work internally. 

I'm all for using `bind` if that is the best way to tackle it 👍🏻

Reply:
Yes, it's reactive to `state` and `context`, but it won't react (rerender) because we are not accessing them, so it will only run once.

Reply:
If I am not mistaken, it should run only once as it has no references that change. I run a quick test and that part seems to work.

However, **it seems the negation operator doesn't work with selectors**. My first guess is that in [these functions](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/utils/interactivity/hooks.js#L21-L42), we are not considering properly that `current` can be a function. I believe this could be solved moving the `hasNegationOperator` logic to the `getEvaluate` function. Something like this:

```js
// Resolve the path to some property of the store object.
const resolve = ( path, ctx ) => {
	let current = { ...store, context: ctx };
	path.split( '.' ).forEach( ( p ) => ( current = current[ p ] ) );
	return current;
};

// Generate the evaluate function.
const getEvaluate =
	( { ref } = {} ) =>
	( path, extraArgs = {} ) => {
		// If path starts with !, remove it and save a flag.
		const hasNegationOperator =
			path[ 0 ] === '!' && !! ( path = path.slice( 1 ) );
		const value = resolve( path, extraArgs.context );

		const returnValue =
			typeof value === 'function'
				? value( {
						ref: ref.current,
						...store,
						...extraArgs,
				  } )
				: value;

		return hasNegationOperator ? ! returnValue : returnValue;
	};
```

I quickly tested and it seems to work, but I am not sure if it is the correct approach.

Reply:
I will wrap the function for now in the custom selector to handle negation, but we should follow-up with extending the handling for `!` as it seems to important to have it working universally with both the context and selectors.

Reply:
It works correctly with https://github.com/WordPress/gutenberg/pull/50377/commits/2d1c63b7a89c8a89e9bb9924c92e04cc0592558a 🎉

Reply:
I tested that the `bind` never updates again with the following code:

```diff
diff --git a/lib/experimental/interactivity-api/blocks.php b/lib/experimental/interactivity-api/blocks.php
index 755c1d1d4f..8f900501f6 100644
--- a/lib/experimental/interactivity-api/blocks.php
+++ b/lib/experimental/interactivity-api/blocks.php
@@ -24,6 +24,7 @@ function gutenberg_block_core_file_add_directives_to_content( $block_content, $b
 	$processor->set_attribute( 'data-wp-island', '' );
 	$processor->next_tag( 'object' );
 	$processor->set_attribute( 'data-wp-bind.hidden', 'selectors.core.file.hasNoPdfPreview' );
+	$processor->set_attribute( 'data-wp-on.click', 'actions.core.file.incrementCount' );
 	$processor->set_attribute( 'hidden', true );
 	return $processor->get_updated_html();
 }
diff --git a/packages/block-library/src/file/interactivity.js b/packages/block-library/src/file/interactivity.js
index cf9ae41002..76696511cc 100644
--- a/packages/block-library/src/file/interactivity.js
+++ b/packages/block-library/src/file/interactivity.js
@@ -5,13 +5,27 @@ import { store } from '../utils/interactivity';
 import { browserSupportsPdfs } from './utils';
 
 store( {
+	state: {
+		count: 0,
+	},
 	selectors: {
 		core: {
 			file: {
 				hasNoPdfPreview() {
+					console.log( 'Run!' );
 					return ! browserSupportsPdfs();
 				},
 			},
 		},
 	},
+	actions: {
+		core: {
+			file: {
+				incrementCount( { state } ) {
+					state.count = state.count + 1;
+					console.log( state.count );
+				},
+			},
+		},
+	},
 } );
```

This is what I see printed in the console when clicking on the PDF preview:

<img width="690" alt="Screenshot 2023-05-17 at 10 18 26" src="https://github.com/WordPress/gutenberg/assets/699132/e27f9129-7adc-4fbe-b91c-7e93b9444d4a">

---

It made me realize that we use Preact Signals behind the scenes, which means that the selector doesn't use any signal, so it will never update as there is nothing it needs to listen to for change ...

As simple as that when you switch the mindset from React 😄

Reply:
> I believe this could be solved moving the hasNegationOperator logic to the getEvaluate function.

That seems much better, yes 🙂

Reply:
I created a [pull request in the block-interactivity-experiments](https://github.com/WordPress/block-interactivity-experiments/pull/232) repo to handle that. We can keep the conversation about that there.